### ISSUE TITLE
socket.cr: Close on exec on sockets does not work on platforms that supports LibC::SOCK_CLOEXEC

### ISSUE THREAD

**Issue by DRVTiny** [crystal-lang/crystal#8696]
**init_close_on_exec**  method of the **Socket** class will do nothing if the current/target platform supports specifying close_on_exec option as a part of bitmask in the "type" argument of LibC.socket() :

The problem is in the code of init_close_on_exec in master's socket.cr:
```
  protected def init_close_on_exec(fd : Int32)
    {% unless LibC.has_constant?(:SOCK_CLOEXEC) %}
      LibC.fcntl(fd, LibC::F_SETFD, LibC::FD_CLOEXEC)
    {% end %}
  end
```
- here close_on_exec option will be set via fcntl() call only in cases where socket() does not support  this option in time of fd creation.

I propose to remove macro-if from init_close_on_exec:
```
  protected def init_close_on_exec(fd : Int32)
    LibC.fcntl(fd, LibC::F_SETFD, LibC::FD_CLOEXEC)
  end
```

And modify first constructor in such manner:
```
  def initialize(@family, @type, @protocol = Protocol::IP, blocking = false)
    @closed = false
    fd = {% begin %}LibC.socket(
      family,
      type.value{% if LibC.has_constant?(:SOCK_CLOEXEC) %} | LibC::SOCK_CLOEXEC{% end %},
      protocol
    )
    {% end %}
    raise Errno.new("failed to create socket:") if fd == -1
    
    {% unless LibC.has_constant?(:SOCK_CLOEXEC) %}
      init_close_on_exec(fd)
    {% end %}
        
    @fd = fd    
```

So in the second Socket's constructor init_close_on_exec(fd) will do its work properly in all cases:
```
  def initialize(@fd : Int32, @family, @type, @protocol = Protocol::IP, blocking = false)
    @closed = false
    init_close_on_exec(@fd)
```

Thank you!

**Comment by DRVTiny** [comment]
According to fcntl man pages, setting FD_CLOEXEC not inside system call intended to create new file descriptor (such as open() or socket()) - may be dangerous in multithreaded environment because of possible race conditions: if one thread set fcntl flag FD_CLOEXEC before another thread fork/execve new process - file descriptor will be closed in this (new) process, and vice versa: another thread may fork() before our thread set fcntl, so there will be unnecessary opened file desriptors. This can be avoided using open using open system call with O_CLOEXEC flag:
( http://man7.org/linux/man-pages/man2/open.2.html ) 

> O_CLOEXEC (since Linux 2.6.23)
>               Enable the close-on-exec flag for the new file descriptor.
>               Specifying this flag permits a program to avoid additional
>               fcntl(2) F_SETFD operations to set the FD_CLOEXEC flag.
> 
>               Note that the use of this flag is essential in some
>               multithreaded programs, because using a separate fcntl(2)
>               F_SETFD operation to set the FD_CLOEXEC flag does not suffice
>               to avoid race conditions where one thread opens a file
>               descriptor and attempts to set its close-on-exec flag using
>               fcntl(2) at the same time as another thread does a fork(2)
>               plus execve(2).  Depending on the order of execution, the race
>               may lead to the file descriptor returned by open() being
>               unintentionally leaked to the program executed by the child
>               process created by fork(2).  (This kind of race is in
>               principle possible for any system call that creates a file
>               descriptor whose close-on-exec flag should be set, and various
>               other Linux system calls provide an equivalent of the
>               O_CLOEXEC flag to deal with this problem.)

But in my opinion we can:
1) Disable any Crystal runtime public methods operating on file descriptor numbers directly (only on IO::FileDescriptor)
2) Close manually all opened file descriptors by default after Process::fork(), Process::new(), but let the user to define explicitly, what fds needs to be stay opened after fork.

I am not sure, but guess that 2 - can be possible safely only in single-threaded configuration of the forked process with temporary disabled event loop.

**Comment by ysbaddaden** [comment]
Better late than never: this was fixed in #14672, #14673 and #14674.
